# Bigcontest 데이터 탐색

## 0. library Packages

In [1]:
import pandas as pd
import numpy as np

## 1. Import Data

In [2]:
loan_result = pd.read_csv("../data/loan_result.csv")
log_data = pd.read_csv("../data/log_data.csv")
user_spec = pd.read_csv("../data/user_spec_new.csv")

In [4]:
user_spec.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1384403 entries, 0 to 1384402
Data columns (total 18 columns):
 #   Column                               Non-Null Count    Dtype  
---  ------                               --------------    -----  
 0   Unnamed: 0                           1384403 non-null  int64  
 1   application_id                       1384403 non-null  int64  
 2   user_id                              1384403 non-null  int64  
 3   insert_time                          1384403 non-null  object 
 4   credit_score                         1280926 non-null  float64
 5   yearly_income                        1384403 non-null  float64
 6   income_type                          1384403 non-null  object 
 7   company_enter_month                  1384403 non-null  int64  
 8   employment_type                      1384403 non-null  object 
 9   houseown_type                        1384403 non-null  object 
 10  desired_amount                       1384403 non-null  float64
 11

In [13]:
merged = pd.merge(user_spec, train_result, on = 'application_id', how = 'inner')

In [14]:
merged

Unnamed: 0  application_id  user_id          insert_time  \
0                154         1945260   836762  2022-04-20 01:23:09   
1                154         1945260   836762  2022-04-20 01:23:09   
2                157         1019382   186886  2022-04-20 00:38:18   
3                157         1019382   186886  2022-04-20 00:38:18   
4                157         1019382   186886  2022-04-20 00:38:18   
...              ...             ...      ...                  ...   
10192983     1384492          816537     3864  2022-03-22 08:55:14   
10192984     1384492          816537     3864  2022-03-22 08:55:14   
10192985     1384492          816537     3864  2022-03-22 08:55:14   
10192986     1384492          816537     3864  2022-03-22 08:55:14   
10192987     1384492          816537     3864  2022-03-22 08:55:14   

          credit_score  yearly_income    income_type  company_enter_month  \
0                670.0     50000000.0  EARNEDINCOME2               202110   
1                670.0     50000000.0  EARNEDINCOME2               202110   
2                730.0     95000000.0   EARNEDINCOME               201803   
3                730.0     95000000.0   EARNEDINCOME               201803   
4                730.0     95000000.0   EARNEDINCOME               201803   
...                ...            ...            ...                  ...   
10192983           NaN     35000000.0     FREELANCER               201103   
10192984           NaN     35000000.0     FREELANCER               201103   
10192985           NaN     35000000.0     FREELANCER               201103   
10192986           NaN     35000000.0     FREELANCER               201103   
10192987           NaN     35000000.0     FREELANCER               201103   

         employment_type houseown_type  ...  existing_loan_cnt  \
0                     기타           전월세  ...                  3   
1                     기타           전월세  ...                  3   
2                    정규직           전월세  ...                  4   
3                    정규직           전월세  ...                  4   
4                    정규직           전월세  ...                  4   
...                  ...           ...  ...                ...   
10192983              기타            자가  ...                  0   
10192984              기타            자가  ...                  0   
10192985              기타            자가  ...                  0   
10192986              기타            자가  ...                  0   
10192987              기타            자가  ...                  0   

         existing_loan_amt  birth_year  gender  loanapply_insert_time  \
0               76000000.0        1979       1    2022-04-20 01:23:10   
1               76000000.0        1979       1    2022-04-20 01:23:09   
2              142000000.0        1979       1    2022-04-20 00:38:20   
3              142000000.0        1979       1    2022-04-20 00:38:19   
4              142000000.0        1979       1    2022-04-20 00:38:20   
...                    ...         ...     ...                    ...   
10192983               NaN        1977       0    2022-03-22 08:55:14   
10192984               NaN        1977       0    2022-03-22 08:55:14   
10192985               NaN        1977       0    2022-03-22 08:55:15   
10192986               NaN        1977       0    2022-03-22 08:55:15   
10192987               NaN        1977       0    2022-03-22 08:55:16   

          bank_id  product_id  loan_limit loan_rate  is_applied  
0              42         216   3000000.0      14.5         0.0  
1              13         123   1000000.0      19.9         0.0  
2              15         141  11000000.0      15.1         0.0  
3               1          61  15000000.0       9.9         0.0  
4              49          39   3000000.0      15.9         0.0  
...           ...         ...         ...       ...         ...  
10192983       13         123   8000000.0      20.0         0.0  
10192984       19         231  50000000.0     

In [12]:
user_spec

Unnamed: 0  application_id  user_id          insert_time  \
0                 1         1249046   118218  2022-06-07 06:28:18   
1                 2          954900   553686  2022-06-07 14:29:03   
2                 3          137274    59516  2022-06-07 21:40:22   
3                 4         1570936   167320  2022-06-07 09:40:27   
4                 5          967833    33400  2022-06-07 08:55:07   
...             ...             ...      ...                  ...   
1384398     1384488         1864587   489900  2022-03-22 14:55:32   
1384399     1384489         1327066   151422  2022-03-22 01:19:24   
1384400     1384490         1319606   173524  2022-03-22 07:34:32   
1384401     1384491         1482466   766546  2022-03-22 22:12:35   
1384402     1384492          816537     3864  2022-03-22 08:55:14   

         credit_score  yearly_income      income_type  company_enter_month  \
0               660.0    108000000.0  PRIVATEBUSINESS               201511   
1               870.0     30000000.0  PRIVATEBUSINESS               200702   
2               710.0     30000000.0       FREELANCER               202109   
3               820.0     62000000.0     EARNEDINCOME               201701   
4               630.0     36000000.0     EARNEDINCOME               202109   
...               ...            ...              ...                  ...   
1384398         590.0     25000000.0       FREELANCER               202106   
1384399         980.0     20000000.0      OTHERINCOME                    0   
1384400         750.0     75000000.0     EARNEDINCOME               200908   
1384401         640.0     50000000.0     EARNEDINCOME               201705   
1384402           NaN     35000000.0       FREELANCER               201103   

        employment_type houseown_type  desired_amount purpose  \
0                    기타            자가       1000000.0      기타   
1                   정규직        기타가족소유      30000000.0    대환대출   
2                    기타        기타가족소유      10000000.0     생활비   
3                   정규직            자가       2000000.0     생활비   
4                   정규직        기타가족소유       5000000.0     생활비   
...                 ...           ...             ...     ...   
1384398              기타        기타가족소유       5000000.0    사업자금   
1384399              기타            자가      50000000.0     생활비   
1384400             정규직            자가     100000000.0    대환대출   
1384401             정규직            자가      10000000.0    대환대출   
1384402              기타            자가      20000000.0     생활비   

         personal_rehabilitation_yn  personal_rehabilitation_complete_yn  \
0                               0.0                                  NaN   
1                               0.0                                  NaN   
2                               0.0                                  NaN   
3                               0.0                                  NaN   
4                               0.0                                  0.0   
...                             ...                                  ...   
1384398                         NaN                                  NaN   
1384399                         NaN                                  NaN   
1384400                         NaN                                  NaN   
1384401                         NaN                                  NaN   
1384402                         NaN                                  NaN   

         existing_loan_cnt  existing_loan_amt  birth_year  gender  
0                        4        162000000.0        1985       1  
1                        1         27000000.0        1968       1  
2                        5         15000000.0        1997       1  
3                        7        344000000.0        1989       1  
4                        1         16000000.0        2000       1  
...                    ...                ...         ...     ...  
1384398                  0                NaN        2000       1  
13843

In [7]:
loan_result[loan_result['application_id'] == 191834]

application_id loanapply_insert_time  bank_id  product_id  \
13526083          191834   2022-06-03 14:20:53       35          29   
13526084          191834   2022-06-03 14:20:42       59         150   
13526085          191834   2022-06-03 14:20:43       35         168   
13526086          191834   2022-06-03 14:20:44       58         175   
13526087          191834   2022-06-03 14:20:43        6          36   

          loan_limit  loan_rate  is_applied  
13526083   2000000.0       19.4         NaN  
13526084  22000000.0       16.4         NaN  
13526085  13000000.0       19.4         NaN  
13526086  16000000.0       13.8         NaN  
13526087  15000000.0       17.7         NaN

## 2. Preview & Preprocessing

### 2-1) loan_result

![loan_result](../img/loan_result_describe.png)

In [4]:
loan_result

application_id loanapply_insert_time  bank_id  product_id  \
0                1748340   2022-06-07 13:05:41        7         191   
1                1748340   2022-06-07 13:05:41       25         169   
2                1748340   2022-06-07 13:05:41        2           7   
3                1748340   2022-06-07 13:05:41        4         268   
4                1748340   2022-06-07 13:05:41       11         118   
...                  ...                   ...      ...         ...   
13527358         1428218   2022-06-03 12:01:58       62         200   
13527359         1428218   2022-06-03 12:01:49        2           7   
13527360         1428218   2022-06-03 12:01:48       32         257   
13527361         1428218   2022-06-03 12:01:48       33         110   
13527362         1428218   2022-06-03 12:01:48        5         194   

          loan_limit  loan_rate  is_applied  
0         42000000.0       13.6         NaN  
1         24000000.0       17.9         NaN  
2         24000000.0       18.5         NaN  
3         29000000.0       10.8         NaN  
4          5000000.0       16.4         NaN  
...              ...        ...         ...  
13527358   3000000.0       14.8         NaN  
13527359  40000000.0       11.8         NaN  
13527360  15000000.0        7.2         NaN  
13527361  44000000.0       13.5         NaN  
13527362  44000000.0        9.7         NaN  

[13527363 rows x 7 columns]

13,527,363 행과 7열로 데이터의 행 개수가 많음

In [5]:
loan_result = loan_result.astype({"loanapply_insert_time":'datetime64'})

한도조회 일시를 나타내는 `loanapply_insert_time` 변수가 object 형태로 저장되어 있어 `datetime64`로 정리함.

In [5]:
loan_result.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13527363 entries, 0 to 13527362
Data columns (total 7 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   application_id         int64         
 1   loanapply_insert_time  datetime64[ns]
 2   bank_id                int64         
 3   product_id             int64         
 4   loan_limit             float64       
 5   loan_rate              float64       
 6   is_applied             float64       
dtypes: datetime64[ns](1), float64(3), int64(3)
memory usage: 722.4 MB


In [6]:
loan_result.describe(include = ['float64'])

loan_limit     loan_rate    is_applied
count  1.351987e+07  1.351987e+07  1.027012e+07
mean   2.274095e+07  1.275172e+01  5.457081e-02
std    7.221901e+07  3.843386e+00  2.271406e-01
min    0.000000e+00  1.500000e+00  0.000000e+00
25%    9.000000e+06  9.600000e+00  0.000000e+00
50%    1.800000e+07  1.320000e+01  0.000000e+00
75%    3.300000e+07  1.560000e+01  0.000000e+00
max    2.308000e+11  2.000000e+01  1.000000e+00

**대출 승인(loan_limit)** 한도의 최댓값은 2천 308억원이며, 75% 값은 3천 300만원, 50% 값은 1천 800만원, 25% 값은 9백만원이다.  
**승인 금리(loan_rate)** 의 최댓값은 20, 75% 값은 15.6, 50% 값은 13.2, 25% 값은 9.6이다.  
**신청 여부(is_applied)** 변수에 NaN이 아닌 값의 개수는 10,270,124개이다. NaN 개수는 3,257,239개이며, 이는 대략 전체 데이터의 24.1% 이다.

In [7]:
print(loan_result["is_applied"].value_counts())
print(loan_result["is_applied"].value_counts(normalize=True))

0.0    9709675
1.0     560449
Name: is_applied, dtype: int64
0.0    0.945429
1.0    0.054571
Name: is_applied, dtype: float64


NaN 값이 아닌 **신청 여부(is_applied)** 값들 중 대략 94.5%가 신청을 하지 않았고, 5.5%가 신청을 하였다.  
1의 비율이 적으므로 이를 고려할 수 있는 모델을 만드는 것이 중요해보인다.

### 2-2) log_data

![log_data](../img/log_data_describe.png)

In [8]:
log_data

user_id               event            timestamp    mp_os  \
0          576409      StartLoanApply  2022-03-25 11:12:09  Android   
1          576409  ViewLoanApplyIntro  2022-03-25 11:12:09  Android   
2           72878        EndLoanApply  2022-03-25 11:14:44  Android   
3          645317             OpenApp  2022-03-25 11:15:09      iOS   
4          645317       UseLoanManage  2022-03-25 11:15:11      iOS   
...           ...                 ...                  ...      ...   
17843988   242442      StartLoanApply  2022-05-27 16:08:21  Android   
17843989   242442        EndLoanApply  2022-05-27 16:08:21  Android   
17843990   593062             OpenApp  2022-05-27 16:19:10  Android   
17843991   593062               Login  2022-05-27 16:19:15  Android   
17843992   593062       UseLoanManage  2022-05-27 16:19:19  Android   

         mp_app_version     date_cd  
0                 3.8.2  2022-03-25  
1                 3.8.2  2022-03-25  
2                 3.8.4  2022-03-25  
3                 3.6.1  2022-03-25  
4                 3.6.1  2022-03-25  
...                 ...         ...  
17843988         3.13.0  2022-05-27  
17843989         3.13.0  2022-05-27  
17843990         3.13.0  2022-05-27  
17843991         3.13.0  2022-05-27  
17843992         3.13.0  2022-05-27  

[17843993 rows x 6 columns]

17,843,993 행과 6열로 데이터의 행 개수가 많음

In [9]:
log_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17843993 entries, 0 to 17843992
Data columns (total 6 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   user_id         int64 
 1   event           object
 2   timestamp       object
 3   mp_os           object
 4   mp_app_version  object
 5   date_cd         object
dtypes: int64(1), object(5)
memory usage: 816.8+ MB


In [11]:
log_data = log_data.drop(labels = ["mp_os", "mp_app_version","date_cd"], axis = 1).astype({'timestamp':'datetime64'})

행동일시를 나타내는 `timestamp` 가 object 형태이므로 `datetime64`로 정리할 필요가 있어보임. 그리고, 사용 기종과 어플리케이션 버전이 크게 유의미하지 않을 것이라 생각되고, `data_cd`는 `timestamp`에 포함되므로 삭제한다.

In [12]:
log_data["event"].value_counts()

OpenApp                    3460762
EndLoanApply               2715253
GetCreditInfo              2661997
Login                      2463755
StartLoanApply             1893914
ViewLoanApplyIntro         1804712
UseLoanManage              1558906
CompleteIDCertification    1237777
SignUp                       34892
UsePrepayCalc                 7360
UseDSRCalc                    4665
Name: event, dtype: int64

### 2-3) user_spec

![user_spec](../img/user_spec_describe.png)

In [13]:
user_spec

application_id  user_id  birth_year  gender          insert_time  \
0               1249046   118218      1985.0     1.0  2022-06-07 06:28:18   
1                954900   553686      1968.0     1.0  2022-06-07 14:29:03   
2                137274    59516      1997.0     1.0  2022-06-07 21:40:22   
3               1570936   167320      1989.0     1.0  2022-06-07 09:40:27   
4                967833    33400      2000.0     1.0  2022-06-07 08:55:07   
...                 ...      ...         ...     ...                  ...   
1394211         1864587   489900      2000.0     1.0  2022-03-22 14:55:32   
1394212         1327066   151422      1955.0     1.0  2022-03-22 01:19:24   
1394213         1319606   173524      1983.0     1.0  2022-03-22 07:34:32   
1394214         1482466   766546      1975.0     1.0  2022-03-22 22:12:35   
1394215          816537     3864      1977.0     0.0  2022-03-22 08:55:14   

         credit_score  yearly_income      income_type  company_enter_month  \
0               660.0    108000000.0  PRIVATEBUSINESS           20151101.0   
1               870.0     30000000.0  PRIVATEBUSINESS           20070201.0   
2               710.0     30000000.0       FREELANCER           20210901.0   
3               820.0     62000000.0     EARNEDINCOME           20170101.0   
4               630.0     36000000.0     EARNEDINCOME           20210901.0   
...               ...            ...              ...                  ...   
1394211         590.0     25000000.0       FREELANCER             202106.0   
1394212         980.0     20000000.0      OTHERINCOME                  NaN   
1394213         750.0     75000000.0     EARNEDINCOME             200908.0   
1394214         640.0     50000000.0     EARNEDINCOME             201705.0   
1394215           NaN     35000000.0       FREELANCER             201103.0   

        employment_type houseown_type  desired_amount purpose  \
0                    기타            자가       1000000.0      기타   
1                   정규직        기타가족소유      30000000.0    대환대출   
2                    기타        기타가족소유      10000000.0     생활비   
3                   정규직            자가       2000000.0     생활비   
4                   정규직        기타가족소유       5000000.0     생활비   
...                 ...           ...             ...     ...   
1394211              기타        기타가족소유       5000000.0    사업자금   
1394212              기타            자가      50000000.0     생활비   
1394213             정규직            자가     100000000.0    대환대출   
1394214             정규직            자가      10000000.0    대환대출   
1394215              기타            자가      20000000.0     생활비   

         personal_rehabilitation_yn  personal_rehabilitation_complete_yn  \
0                               0.0                                  NaN   
1                               0.0                                  NaN   
2                               0.0                                  NaN   
3                               0.0                                  NaN   
4                               0.0                                  0.0   
...                             ...                                  ...   
1394211                         NaN                                  NaN   
1394212                         NaN                                  NaN   
1394213                         NaN                                  NaN   
1394214                         NaN                                  NaN   
1394215                         NaN                                  NaN   

         existing_loan_cnt  existing_loan_amt  
0                      4.0        162000000.0  
1                      1.0         27000000.0  
2                      5.0         15000000.0  
3                      7.0        344000000.0  
4                      1.0         16000000.0  
...                    ...                ...  
1394211                NaN                NaN  
1394212                1.0                NaN  
1394213              

1,394,216 행과 17열로 각 행이 한 사람에 대한 정보를 뜻함

In [14]:
user_spec = user_spec.astype({user_spec.columns[2]:'Int64', user_spec.columns[3]:'Int64', user_spec.columns[4]:'datetime64', 
                              user_spec.columns[5]:'Int64', user_spec.columns[6]:'Int64', user_spec.columns[11]:'Int64'})

In [15]:
def convert_time(x):
    if np.isnan(x):
        return x
    if np.log10(x)>7:
        x = x // 100
    x = pd.to_datetime(str(int(x)), format='%Y%m')
    return x

In [16]:
foo = user_spec["company_enter_month"]
foo = foo.map(convert_time)
user_spec["company_enter_month"] = foo

적절히 `dtype`을 변경하고, 시간을 나타내는 변수들에 대해서 적당히 `datetime64`로 변경한다.

In [17]:
user_spec.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1394216 entries, 0 to 1394215
Data columns (total 17 columns):
 #   Column                               Non-Null Count    Dtype         
---  ------                               --------------    -----         
 0   application_id                       1394216 non-null  int64         
 1   user_id                              1394216 non-null  int64         
 2   birth_year                           1381255 non-null  Int64         
 3   gender                               1381255 non-null  Int64         
 4   insert_time                          1394216 non-null  datetime64[ns]
 5   credit_score                         1289101 non-null  Int64         
 6   yearly_income                        1394126 non-null  Int64         
 7   income_type                          1394131 non-null  object        
 8   company_enter_month                  1222456 non-null  datetime64[ns]
 9   employment_type                      1394131 non-null  ob

In [18]:
user_spec.isnull().sum()

application_id                               0
user_id                                      0
birth_year                               12961
gender                                   12961
insert_time                                  0
credit_score                            105115
yearly_income                               90
income_type                                 85
company_enter_month                     171760
employment_type                             85
houseown_type                               85
desired_amount                              85
purpose                                     85
personal_rehabilitation_yn              587461
personal_rehabilitation_complete_yn    1203354
existing_loan_cnt                       198556
existing_loan_amt                       313774
dtype: int64

**id** 를 나타내는 변수와  `insert_time` 변수를 제외하고는 NaN이 포함되어 있음.  
이를 확인해보고 결측치 대체나 제거를 고려해야함

## 3. Handling missing data

### 3-1) user_spec

`loan_result`에 있는 `application_id`에 포함되는 `user_spec`만 추출

In [19]:
id_list = loan_result['application_id'].unique()

In [21]:
user_spec2 = user_spec.loc[user_spec['application_id'].isin(id_list),]
user_spec2

application_id  user_id  birth_year  gender         insert_time  \
1                954900   553686        1968       1 2022-06-07 14:29:03   
2                137274    59516        1997       1 2022-06-07 21:40:22   
3               1570936   167320        1989       1 2022-06-07 09:40:27   
4                967833    33400        2000       1 2022-06-07 08:55:07   
6               1614045   860410        1976       1 2022-06-07 11:29:05   
...                 ...      ...         ...     ...                 ...   
1394211         1864587   489900        2000       1 2022-03-22 14:55:32   
1394212         1327066   151422        1955       1 2022-03-22 01:19:24   
1394213         1319606   173524        1983       1 2022-03-22 07:34:32   
1394214         1482466   766546        1975       1 2022-03-22 22:12:35   
1394215          816537     3864        1977       0 2022-03-22 08:55:14   

         credit_score  yearly_income      income_type company_enter_month  \
1                 870       30000000  PRIVATEBUSINESS          2007-02-01   
2                 710       30000000       FREELANCER          2021-09-01   
3                 820       62000000     EARNEDINCOME          2017-01-01   
4                 630       36000000     EARNEDINCOME          2021-09-01   
6                 590       45000000      OTHERINCOME                 NaT   
...               ...            ...              ...                 ...   
1394211           590       25000000       FREELANCER          2021-06-01   
1394212           980       20000000      OTHERINCOME                 NaT   
1394213           750       75000000     EARNEDINCOME          2009-08-01   
1394214           640       50000000     EARNEDINCOME          2017-05-01   
1394215          <NA>       35000000       FREELANCER          2011-03-01   

        employment_type houseown_type  desired_amount purpose  \
1                   정규직        기타가족소유        30000000    대환대출   
2                    기타        기타가족소유        10000000     생활비   
3                   정규직            자가         2000000     생활비   
4                   정규직        기타가족소유         5000000     생활비   
6                    기타            자가        15000000    대환대출   
...                 ...           ...             ...     ...   
1394211              기타        기타가족소유         5000000    사업자금   
1394212              기타            자가        50000000     생활비   
1394213             정규직            자가       100000000    대환대출   
1394214             정규직            자가        10000000    대환대출   
1394215              기타            자가        20000000     생활비   

         personal_rehabilitation_yn  personal_rehabilitation_complete_yn  \
1                               0.0                                  NaN   
2                               0.0                                  NaN   
3                               0.0                                  NaN   
4                               0.0                                  0.0   
6                               0.0                                  NaN   
...                             ...                                  ...   
1394211                         NaN                                  NaN   
1394212                         NaN                                  NaN   
1394213                         NaN                                  NaN   
1394214                         NaN                                  NaN   
1394215                         NaN                                  NaN   

         existing_loan_cnt  existing_loan_amt  
1                      1.0         27000000.0  
2                      5.0         15000000.0  
3                      7.0        344000000.0  
4                      1.0         16000000.0  
6                      1.0         20000000.0  
...                    ...                ...  
1394211                NaN                NaN  
1394212                1.0                NaN  
1394213                8.0        200000000.0

`user_spec2`에 있는 생년과 성별 처리

In [83]:
user_info = user_spec2.groupby('user_id')[['birth_year', 'gender']].mean()
user_info[(user_info['gender'] == 1.0) & (user_info['gender'] == 0.0)] # 성별이 달라지거나 생년이 달라지는 사람 없음!

Empty DataFrame
Columns: [birth_year, gender]
Index: []

In [79]:
user_info

birth_year  gender
user_id                    
9            1996.0     1.0
11           1997.0     1.0
14           1965.0     0.0
17           1993.0     1.0
19           1987.0     0.0
...             ...     ...
879692       1992.0     1.0
879693       2001.0     0.0
879695       1996.0     1.0
879696       1997.0     1.0
879697       1962.0     1.0

[317467 rows x 2 columns]

In [62]:
user_info.isnull().sum() # 아예 생년과 성별이 없는 사람들 확인

birth_year    4954
gender        4954
dtype: int64

In [63]:
user_info = user_info.dropna(axis = 0)
user_info.isnull().sum() # 생년과 성별 없는 사람들 아예 제거

birth_year    0
gender        0
dtype: int64

In [64]:
user_info

birth_year  gender
user_id                    
9            1996.0     1.0
11           1997.0     1.0
14           1965.0     0.0
17           1993.0     1.0
19           1987.0     0.0
...             ...     ...
879692       1992.0     1.0
879693       2001.0     0.0
879695       1996.0     1.0
879696       1997.0     1.0
879697       1962.0     1.0

[312513 rows x 2 columns]

In [113]:
user_spec3 = user_spec2.loc[user_spec['user_id'].isin(user_info.index),] 

In [115]:
user_spec3.isnull().sum()

application_id                              0
user_id                                     0
birth_year                               8593
gender                                   8593
insert_time                                 0
credit_score                            81769
yearly_income                               1
income_type                                 0
company_enter_month                     92314
employment_type                             0
houseown_type                               0
desired_amount                              0
purpose                                     0
personal_rehabilitation_yn             417763
personal_rehabilitation_complete_yn    843993
existing_loan_cnt                      146290
existing_loan_amt                      225046
dtype: int64

In [1]:
na_list = user_spec3[user_spec3['birth_year'].isnull()]['user_id'].unique()
for user in na_list:
    user_spec3.loc[user_spec3['user_id'] == user,:].loc[:,['birth_year','gender']].fillna(user_info.loc[user,:], inplace = True)
    

NameError: name 'user_spec3' is not defined

In [146]:
user_spec3.loc[user_spec3['user_id'] == 49072,].loc[:,['birth_year','gender']].fillna(user_info.loc[49072,:])

birth_year  gender
253           1985       0
266825        1985       0
405894        1985       0
417856        1985       0
420503        1985       0
421707        1985       0
673697        1985       0
816204        1985       0

In [150]:
user_spec3.loc[user_spec3['user_id'] == 49072,].loc[:,['birth_year','gender']]

birth_year  gender
253           <NA>    <NA>
266825        1985       0
405894        1985       0
417856        1985       0
420503        1985       0
421707        1985       0
673697        1985       0
816204        1985       0

In [140]:
user_spec3.isnull().sum()

application_id                              0
user_id                                     0
birth_year                               8593
gender                                   8593
insert_time                                 0
credit_score                            81769
yearly_income                               1
income_type                                 0
company_enter_month                     92314
employment_type                             0
houseown_type                               0
desired_amount                              0
purpose                                     0
personal_rehabilitation_yn             417763
personal_rehabilitation_complete_yn    843993
existing_loan_cnt                      146290
existing_loan_amt                      225046
dtype: int64

In [109]:
user_spec3[user_spec3['user_id'] == 49072][['birth_year','gender']].fillna(user_info.loc[49072,])

birth_year  gender
253           1985       0
266825        1985       0
405894        1985       0
417856        1985       0
420503        1985       0
421707        1985       0
673697        1985       0
816204        1985       0

그 다음 분석에 쓰일 데이터와 테스트 데이터를 나눔

In [89]:
user_spec2.shape

(968753, 17)

In [87]:
user_spec3.shape

(961862, 17)

In [ ]:
loan_result_data = loan_result.loc[-loan_result["is_applied"].isna(),].reset_index()
loan_result_test = loan_result.loc[loan_result["is_applied"].isna(),].reset_index()

In [ ]:
loan_result_data

index  application_id loanapply_insert_time  bank_id  product_id  \
0            13284         2157865   2022-05-09 08:44:59       54         235   
1            13285          576643   2022-05-09 10:54:53       54         235   
2            13286          576643   2022-05-09 10:54:53       11         118   
3            13287         2136706   2022-05-09 10:41:06       42         216   
4            13288         2136706   2022-05-09 10:41:07       25         169   
...            ...             ...                   ...      ...         ...   
10270119  13519634         1969227   2022-05-16 14:42:58        2           7   
10270120  13519635         1969227   2022-05-16 14:42:57       33         110   
10270121  13519636         1969227   2022-05-16 14:42:56       50         142   
10270122  13519637         1969227   2022-05-16 14:43:18       22         100   
10270123  13519638         1969227   2022-05-16 14:42:56       19         231   

          loan_limit  loan_rate  is_applied  year  month  day      time  
0         20000000.0       16.5         1.0  2022      5    9  08:44:59  
1         11000000.0       16.5         0.0  2022      5    9  10:54:53  
2          3000000.0       20.0         0.0  2022      5    9  10:54:53  
3         10000000.0       13.5         0.0  2022      5    9  10:41:06  
4         22000000.0       15.9         0.0  2022      5    9  10:41:07  
...              ...        ...         ...   ...    ...  ...       ...  
10270119  30000000.0       13.6         0.0  2022      5   16  14:42:58  
10270120   9000000.0       14.4         0.0  2022      5   16  14:42:57  
10270121   3000000.0       11.2         0.0  2022      5   16  14:42:56  
10270122   4000000.0       15.3         0.0  2022      5   16  14:43:18  
10270123   9000000.0       15.5         0.0  2022      5   16  14:42:56  

[10270124 rows x 12 columns]

In [ ]:
loan_result_test

index  application_id loanapply_insert_time  bank_id  product_id  \
0               0         1748340   2022-06-07 13:05:41        7         191   
1               1         1748340   2022-06-07 13:05:41       25         169   
2               2         1748340   2022-06-07 13:05:41        2           7   
3               3         1748340   2022-06-07 13:05:41        4         268   
4               4         1748340   2022-06-07 13:05:41       11         118   
...           ...             ...                   ...      ...         ...   
3257234  13527358         1428218   2022-06-03 12:01:58       62         200   
3257235  13527359         1428218   2022-06-03 12:01:49        2           7   
3257236  13527360         1428218   2022-06-03 12:01:48       32         257   
3257237  13527361         1428218   2022-06-03 12:01:48       33         110   
3257238  13527362         1428218   2022-06-03 12:01:48        5         194   

         loan_limit  loan_rate  is_applied  year  month  day      time  
0        42000000.0       13.6         NaN  2022      6    7  13:05:41  
1        24000000.0       17.9         NaN  2022      6    7  13:05:41  
2        24000000.0       18.5         NaN  2022      6    7  13:05:41  
3        29000000.0       10.8         NaN  2022      6    7  13:05:41  
4         5000000.0       16.4         NaN  2022      6    7  13:05:41  
...             ...        ...         ...   ...    ...  ...       ...  
3257234   3000000.0       14.8         NaN  2022      6    3  12:01:58  
3257235  40000000.0       11.8         NaN  2022      6    3  12:01:49  
3257236  15000000.0        7.2         NaN  2022      6    3  12:01:48  
3257237  44000000.0       13.5         NaN  2022      6    3  12:01:48  
3257238  44000000.0        9.7         NaN  2022      6    3  12:01:48  

[3257239 rows x 12 columns]